File: 2021/perl/day2.pl
use strict;
use warnings;

open my $file_h, '<', '../input/day2' or die "Cannot open day2 file: $!";

my $total_forward = 0;
my $total_depth   = 0;
my $part_2_depth  = 0;
my $aim           = 0;
while (my $line = <$file_h>)
{
    chomp $line;
    my ($command, $value) = split(' ', $line);
    if (substr($command, 0, 1) eq 'f')
    {
        $total_forward += $value;
        $part_2_depth  += $aim * $value;
    }
    if (substr($command, 0, 1) eq 'd')
    {
        $total_depth += $value;
        $aim         += $value;
    }
    if (substr($command, 0, 1) eq 'u')
    {
        $total_depth -= $value;
        $aim         -= $value;
    }
}
print $total_forward*$total_depth,  "\n";
print $total_forward*$part_2_depth, "\n";

File: 2021/cpp/day2.cpp
#include <charconv>
#include <fstream>
#include <iostream>
#include <ranges>

template <typename R>
inline int convert_to_int(R&& r)
{
    int temp;
    std::from_chars(r.data(), r.data() + r.size(), temp);
    return temp;
}

int main()
{
    using std::operator""sv;
    std::ifstream input("../input/day2");
    std::string line;
    int total_depth = 0;
    int total_forward = 0;
    int aim = 0;
    int part_2_depth = 0;
    while (std::getline(input, line))
    {
        auto rng = line | std::views::split(" "sv) | std::views::adjacent<2>;
        for (auto const& i: rng)
        {
            if (*std::get<0>(i).begin() == 'f')
            {
                int temp = convert_to_int(std::get<1>(i));
                total_forward += temp;
                part_2_depth += aim * temp;
                continue;
            }
            if (*std::get<0>(i).begin() == 'u')
            {
                aim -= convert_to_int(std::get<1>(i));
                total_depth -= convert_to_int(std::get<1>(i));
                continue;
            }
            if (*std::get<0>(i).begin() == 'd')
            {
                aim += convert_to_int(std::get<1>(i));
                total_depth += convert_to_int(std::get<1>(i));
                continue;
            }
        }
    }
    std::cout << total_forward * total_depth << "\n";
    std::cout << total_forward * part_2_depth << "\n";
}
